### ISSUE TITLE
[Messenger] Cannot determine if a message can be dispatched without attempting to dispatch

### ISSUE THREAD

**Issue by davidbarratt** [symfony/symfony#26964]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Symfony version  | 4.1

If you attempt to dispatch a message that has no handler, you get the following exception:
> No handler for message "App\Entity\RSVP".

Full error message:
```
{
	"message": "No handler for message \"App\\Entity\\RSVP\".",
	"code": 0,
	"exception": "Symfony\\Component\\Messenger\\Exception\\NoHandlerForMessageException",
	"previous": null,
	"file": "\/var\/www\/vendor\/symfony\/messenger\/ContainerHandlerLocator.php",
	"line": 36,
	"trace": [
		"\/var\/www\/vendor\/symfony\/messenger\/Middleware\/HandleMessageMiddleware.php(34): Symfony\\Component\\Messenger\\ContainerHandlerLocator->resolve(Object(App\\Entity\\RSVP))",
		"\/var\/www\/vendor\/symfony\/messenger\/MessageBus.php(56): Symfony\\Component\\Messenger\\Middleware\\HandleMessageMiddleware->handle(Object(App\\Entity\\RSVP), Object(Closure))",
		"\/var\/www\/vendor\/symfony\/messenger\/Asynchronous\/Middleware\/SendMessageMiddleware.php(53): Symfony\\Component\\Messenger\\MessageBus->Symfony\\Component\\Messenger\\{closure}(Object(App\\Entity\\RSVP))",
		"\/var\/www\/vendor\/symfony\/messenger\/MessageBus.php(56): Symfony\\Component\\Messenger\\Asynchronous\\Middleware\\SendMessageMiddleware->handle(Object(App\\Entity\\RSVP), Object(Closure))",
		"\/var\/www\/vendor\/symfony\/messenger\/DataCollector\/MessengerDataCollector.php(64): Symfony\\Component\\Messenger\\MessageBus->Symfony\\Component\\Messenger\\{closure}(Object(App\\Entity\\RSVP))",
		"\/var\/www\/vendor\/symfony\/messenger\/MessageBus.php(56): Symfony\\Component\\Messenger\\DataCollector\\MessengerDataCollector->handle(Object(App\\Entity\\RSVP), Object(Closure))",
		"\/var\/www\/vendor\/symfony\/messenger\/Debug\/LoggingMiddleware.php(40): Symfony\\Component\\Messenger\\MessageBus->Symfony\\Component\\Messenger\\{closure}(Object(App\\Entity\\RSVP))",
		"\/var\/www\/vendor\/symfony\/messenger\/MessageBus.php(56): Symfony\\Component\\Messenger\\Debug\\LoggingMiddleware->handle(Object(App\\Entity\\RSVP), Object(Closure))",
		"\/var\/www\/vendor\/symfony\/messenger\/Middleware\/ValidationMiddleware.php(37): Symfony\\Component\\Messenger\\MessageBus->Symfony\\Component\\Messenger\\{closure}(Object(App\\Entity\\RSVP))",
		"\/var\/www\/vendor\/symfony\/messenger\/MessageBus.php(56): Symfony\\Component\\Messenger\\Middleware\\ValidationMiddleware->handle(Object(App\\Entity\\RSVP), Object(Closure))",
		"\/var\/www\/vendor\/symfony\/messenger\/MessageBus.php(40): Symfony\\Component\\Messenger\\MessageBus->Symfony\\Component\\Messenger\\{closure}(Object(App\\Entity\\RSVP))",
		"\/var\/www\/src\/Controller\/RespondController.php(99): Symfony\\Component\\Messenger\\MessageBus->dispatch(Object(App\\Entity\\RSVP))",
		"\/var\/www\/vendor\/symfony\/http-kernel\/HttpKernel.php(149): App\\Controller\\RespondController->createAction(Object(App\\Entity\\RSVP))",
		"\/var\/www\/vendor\/symfony\/http-kernel\/HttpKernel.php(66): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)",
		"\/var\/www\/vendor\/symfony\/http-kernel\/Kernel.php(190): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)",
		"\/var\/www\/html\/index.php(44): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))",
		"{main}"
	]
}
```

It would be really helpful if `Symfony\Component\Messenger\MessageBusInterface` had a `supports()` method (similar to Symfony Serializer) that would allow you to determine if a given message can be dispatched before attempting to dispatch.
https://github.com/symfony/symfony/blob/4af9003bd6288b5a6043e921edc3465b48068c55/src/Symfony/Component/Messenger/MessageBusInterface.php#L19-L31

**Comment by mvrhov** [comment]
The default MessageBus is behaving alike a CommandBus which means that each message must have handler associated with it. There is a proposal to be able to configure multiple buses with different middlewares, where one of the would behave like EventBus where there is no need to have a handler registered. You can manually define such a bus right now

**Comment by sroze** [comment]
I'm 👎 on adding such thing on the bus. Though, a middleware that catches the `NoHandlerException` exception would do the trick for you. It's a use case covered in the following PR: https://github.com/symfony/symfony/pull/26864

**Comment by sroze** [comment]
Solved #26864 by the introduce of the [`TolerateNoHandler` handler](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Messenger/Middleware/TolerateNoHandler.php) :)
